File: sdks/dotnet/src/Croniq.Runner.Sdk/Protocol/WorkAssignment.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Croniq.Runner.Sdk.Protocol;

/// <summary>
/// One unit of work returned by <c>POST /v1/work/poll</c>. Maps 1:1 to
/// the Rust SDK's <c>WorkAssignment</c> struct.
/// </summary>
internal sealed record WorkAssignment(
    [property: JsonPropertyName("execution_id")] string ExecutionId,
    [property: JsonPropertyName("job_key")] string JobKey,
    [property: JsonPropertyName("fire_at")] string FireAt,
    [property: JsonPropertyName("attempt")] int Attempt,
    [property: JsonPropertyName("metadata")] JsonElement Metadata,
    [property: JsonPropertyName("timeout")] string Timeout,
    // Original logical fire time (RFC 3339); null when the server predates the
    // field. Trailing + defaulted to keep positional construction compatible.
    [property: JsonPropertyName("scheduled_for")] string? ScheduledFor = null);

File: sdks/java/core/src/main/java/io/croniq/runner/protocol/WorkAssignment.java
package io.croniq.runner.protocol;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.JsonNode;

/**
 * One work item delivered in a {@link PollResponse}. {@code metadata} is left
 * as an opaque {@link JsonNode} so handlers can navigate it without the SDK
 * needing to know the job-specific schema.
 *
 * <p>{@code timeout} is a humane string (e.g., {@code "1m"}, {@code "30s"}) —
 * parsed by {@link io.croniq.runner.internal.HumanDuration} at dispatch time.
 *
 * <p>{@code scheduledFor} is the original logical fire time (RFC 3339);
 * {@code null} when the server predates the field — consumers must not fall
 * back to {@code fireAt}.
 */
public record WorkAssignment(
        @JsonProperty("execution_id") String executionId,
        @JsonProperty("job_key") String jobKey,
        @JsonProperty("fire_at") String fireAt,
        @JsonProperty("scheduled_for") String scheduledFor,
        @JsonProperty("attempt") int attempt,
        @JsonProperty("metadata") JsonNode metadata,
        @JsonProperty("timeout") String timeout) {}
